src/jenkins.py

```diff
@@ -705,3 +709,195 @@ def get_agent_name(unit_name: str) -> str:
         The agent node name registered on Jenkins server.
     """
     return unit_name.replace("/", "-")
+
+
+PLUGIN_NAME_GROUP = r"^([a-zA-Z0-9-_]+)"
+WHITESPACE = r"\s*"
+VERSION_GROUP = r"\((.*?)\)"
+DEPENDENCIES_GROUP = r"\[(.*?)\]"
+PLUGIN_CAPTURE = rf"{PLUGIN_NAME_GROUP}{WHITESPACE}{VERSION_GROUP}"
+PLUGIN_LINE_CAPTURE = rf"{PLUGIN_CAPTURE} => {DEPENDENCIES_GROUP}"
+
+
+def _get_plugin_name(plugin_info: str) -> str:
+    """Get plugin name given a plugin info string of format <plugin-shortname> (<plugin-version>).
+
+    Args:
+        plugin_info: Text containing plugin name and plugin version.
+
+    Raises:
+        ValidationError: if the plugin info string does not conform to expected format.
+
+    Returns:
+        The plugin shortname.
+    """
+    match = re.match(PLUGIN_CAPTURE, plugin_info)
+    if not match:
+        raise ValidationError(f"No plugin matched in: {plugin_info}")
+    return match.group(1)
+
+
+def _build_dependencies_lookup(
+    plugin_dependency_outputs: typing.Iterable[str],
+) -> dict[str, tuple[str, ...]]:
+    """Build a lookup table of plugin short name to list of dependency plugin's short names.
+
+    Args:
+        plugin_dependency_outputs: The plugin dependency output from Jenkins Groovy script.
+
+    Returns:
+        The dependency lookup table.
+    """
+    # coverage tool thinks defaultdict() has an exit condition.
+    dependency_lookup: dict[str, tuple[str, ...]] = defaultdict(lambda: ())  # pragma: no cover
+    for line in plugin_dependency_outputs:
+        match = re.match(PLUGIN_LINE_CAPTURE, line)
+        if not match:
+            continue
+        plugin, dependencies = match.group(1), match.group(3)
+        if not dependencies:
+            dependency_lookup[plugin] = ()
+        try:
+            dependency_lookup[plugin] = tuple(
+                _get_plugin_name(dependency) for dependency in dependencies.split(", ")
+            )
+        except ValidationError as exc:
+            logger.error("Invalid plugin dependency, %s", exc)
+            continue
+
+    return dependency_lookup
+
+
+def _set_jenkins_system_message(message: str, container: ops.Container) -> None:
```

Comment:
Do you think we can do a bit more to order these functions logically? I'm surprised this function is listed straight after the `_build_dependencies_lookup` function which hasn't been used yet?

Reply:
Reordered by the order in which they are called. Thank you!